### ISSUE TITLE
Add macro checks to lints

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#48855]
From https://github.com/rust-lang/rust/issues/48385

Many of our lints trigger even when the code is within a macro defined by an external crate. They should include macro checks.

Clippy has [a function](https://github.com/rust-lang-nursery/rust-clippy/blob/a7407f9acc2e216dee23fd5e4fc2ab9b01e4d77b/clippy_lints/src/utils/mod.rs#L67-L95) for this already.

We should include this in librustc::lint somewhere, and use it for various lints. It should _not_ be used for future compat lints.

One possible thing to do here is to work it into the lint framework itself, so that non-future-compat lints which are triggered on macro spans will not actually be emitted.

We should still have the helper function since some lints may involve multiple important spans so they need the ability to check them all.

Ideally there would be a way to turn this off, though -- in case you want to find issues in your macros. It's a tricky problem.

Willing to mentor the basic issue of importing the macro checks and applying them in most lints (or working it into the framework).

**Comment by Manishearth** [comment]
@Dylan-DPC would you like to work on this?

**Comment by oli-obk** [comment]
I'm not sure, but @flip1995  might be already on it (see #48364)

**Comment by Manishearth** [comment]
So that's focusing on span_suggestion -- we need to both fix that and fix the general lint case, _except_ for future compat lints

**Comment by oli-obk** [comment]
In clippy we are doing this on a case-by-case basis. Do we really want to blanket ignore all lints in local expansions of external macros?

**Comment by Manishearth** [comment]
Honestly  I don't know. Future compat lints _must_ show, but aside from that the cases where we want to still show a lint on a macro expansion is when the linted code is completely within the invocation, which can't be detected easily right now.

**Comment by Dylan-DPC** [comment]
@Manishearth Ye sure :+1:

**Comment by Manishearth** [comment]
Any updates?

**Comment by Dylan-DPC** [comment]
@Manishearth is there a common place where all lints are used? or i have to hunt down each lint implementation?

**Comment by Manishearth** [comment]
they all implement `EarlyLintPass` or `LateLintPass`. The implementations are scattered.

What you want to do is tweak `span_lint` and similar functions to do the macro check, except for when the lint is a future compat lint (ignore this part for now)

**Comment by nrc** [comment]
From @cramertj  on Discord: problematic interaction with `#[derive(fail)]`

```
warning: trait objects without an explicit `dyn` are deprecated
   --> src/lib.rs:144:25
    |
144 |         #[derive(Debug, Fail)]
    |                         ^^^^ help: use `dyn`: `dyn (Fail)`
```

**Comment by alexcrichton** [comment]
I've revived https://github.com/rust-lang/rust/pull/49755 at https://github.com/rust-lang/rust/pull/52467, and I've confirmed @nrc that the warning with `derive(Fail)` no longer happens after that PR
